engine/consensus/compliance/engine.go

```diff
@@ -19,98 +17,126 @@ import (
 	"github.com/onflow/flow-go/module/metrics"
 	"github.com/onflow/flow-go/module/trace"
 	"github.com/onflow/flow-go/network"
-	"github.com/onflow/flow-go/state"
 	"github.com/onflow/flow-go/state/protocol"
 	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/fifoqueue"
 	"github.com/onflow/flow-go/utils/logging"
 )
 
-// Engine is the consensus engine, responsible for handling communication for
-// the embedded consensus algorithm.
+type Event struct {
+	OriginID flow.Identifier
+	Msg      interface{}
+}
+
+type (
+	EventSink chan *Event // Channel to push pending events
+)
+
+// defaultBlockQueueCapacity maximum capacity of block proposals queue
+const defaultBlockQueueCapacity = 10000
+
+// defaultVoteQueueCapacity maximum capacity of block votes queue
+const defaultVoteQueueCapacity = 10000
+
+// Engine is a wrapper struct for `Core` which implements consensus algorithm.
+// Engine is responsible for handling incoming messages, queueing for processing, broadcasting proposals.
 type Engine struct {
-	unit     *engine.Unit   // used to control startup/shutdown
-	log      zerolog.Logger // used to log relevant actions with context
-	metrics  module.EngineMetrics
-	tracer   module.Tracer
-	mempool  module.MempoolMetrics
-	spans    module.ConsensusMetrics
-	me       module.Local
-	cleaner  storage.Cleaner
-	headers  storage.Headers
-	payloads storage.Payloads
-	state    protocol.MutableState
-	con      network.Conduit
-	prov     network.Engine
-	pending  module.PendingBlockBuffer // pending block cache
-	sync     module.BlockRequester
-	hotstuff module.HotStuff
+	unit             *engine.Unit
+	log              zerolog.Logger
+	mempool          module.MempoolMetrics
+	metrics          module.EngineMetrics
+	me               module.Local
+	headers          storage.Headers
+	payloads         storage.Payloads
+	tracer           module.Tracer
+	state            protocol.State
+	prov             network.Engine
+	core             *Core
+	pendingEventSink EventSink
+	blockSink        EventSink
+	voteSink         EventSink
+	pendingBlocks    *fifoqueue.FifoQueue
+	pendingVotes     *fifoqueue.FifoQueue
```

Comment:
Talking about tests, it would be better to add tests to the `FifoQueen` module, as we added some functionality on top (WithCapacity), and protect ourselves in case the behavior of the `deque` module changes.

Reply:
there are tests for it